Allow shapecheck of PixelCNN++

juliuskunze opened the pull request:
Makes [shapecheck work on PixelCNN++](https://github.com/JuliusKunze/jaxnet/blob/shapecheck-pixelcnn/tests/test_examples.py#L245). Allows `shapecheck` of indexing, slicing, `broadcast_to`, `reshape`. **Updated:** `np.split`, `np.where` and random sampling are now taken out of this PR for simplification. They required special handling for polymorphic shapes since
```python
@shapecheck(['n'], 'n')
  def neg_zeros_like(x):
    return -np.zeros(x.shape)
```
breaks: `find_top_trace` in `full_p.bind` returns `None` here since all arguments are independent of any inputs. Therefore, the result of `zeros` is not wrapped in a trace, and the `np.negative` _implementation_ is called with a polymorphic shape instead of the shape rule, causing an error. This also breaks shape check of `-np.ones(x.shape)`, `-np.broadcast_to(0, x.shape)`, `-np.arange(10)` etc.

I will keep the workarounds [in a separate branch](https://github.com/JuliusKunze/jax/compare/shapecheck-pcnn...JuliusKunze:shapecheck-pcnn-finish?expand=1) until this issue is fixed.

juliuskunze committed:
Allow shapecheck of indexing, slicing, broadcast_to, reshape, random.uniform, iota, simple cases of split

mattjj commented:
One way to think about this is that `np.zeros`, `np.ones`, `np.arange` etc take no value-level arguments and only take type-level arguments. In NumPy there isn't much of a distinction, but in JAX land we think of shapes as part of the type system.

But as you've seen our `np.zeros` implementation expects to get a tuple of integers, not these new parametrically-polymorphic type arguments that `masking.py` is effectively adding to the system. This also interacts with our tracing mechanism, as you pointed out, which is based on data dependence of values: here `np.zeros(x.shape)` has no data dependence on a value, yet it needs to produce a value with a polymorphic shape (i.e. lifted into the appropriate ShapeCheckTrace instance).

This is tricky to solve, but doable. We had something analogous with `pmap` and things like `lax.psum(1, 'i')` or `lax.axis_index('i')`, where again there's no data dependence on a traced value. Instead we use the fact that the axis name `'i'` corresponds to a trace, and we use that together with a global mapping from axis names to traces (`dynamic_axis_env` in pxla.py) to [lift into the trace](https://github.com/google/jax/blob/28f70cc8f8ac4fc4b98ce85f51b389cec8635704/jax/interpreters/pxla.py#L259).

I think we need to do something similar here: every shape indeterminate is associated with a trace; that information isn't explicitly stored anywhere but you could imagine adapting [`shape_envs`](https://github.com/google/jax/blob/28f70cc8f8ac4fc4b98ce85f51b389cec8635704/jax/interpreters/masking.py#L44) to include this mapping. Then when we call `np.zeros` and notice that we get a polymorphic shape argument, we can use that mapping to lift into the appropriate trace.

Pretty tricky! Maybe we should pair program it :)

WDYT? (I could be wrong about any part of the above!)

JuliusKunze commented:
@mattjj That's a very neat approach and just what we need. In our case, we could simply attach a `ShapeCheckTrace` to `Poly`. This should always be possible since they are currently only created 
- from `ShapeSpec`s, which are passed as arguments into `shapecheck`, where we can access the corresponding `ShapeCheckTrace`, or 
- from existing `Poly`s via operators.

I haven't looked into the `pmap` implementation yet, but I would most certainly need help with the unusual lifting. I guess pair programming might make sense here. (:

Also, it looks like some tests break because `onp.float64` is used for indexing / sizes in other tests. Would that be expected behavior or a bug? This is now an issue I replaced some `int(x)` calls with `x.__index__()`, resulting stricter type requirements. (Will look at this in detail later.)

shoyer commented on jax/random.py:
```diff
@@ -292,7 +300,7 @@ def _random_bits(key, bit_width, shape):
 def _check_shape(name, shape, *param_shapes):
   try:
-    shape = tuple(map(int, shape))
+    shape = tuple(map(lambda x: x.__index__(), shape))
```
The standard way to call `__index__` methods is with operator.index:
```suggestion
    shape = tuple(map(operator.index, shape))
```

JuliusKunze replied:
@shoyer Agree. I tried using `operator.index` originally, but `operator.index(Poly({}))` fails with `TypeError: __index__ returned non-int (type Poly)`. `Poly({}).__index__()` runs fine, however. The suggested change would make `test_shapecheck_uniform` fail.

`operator.index` was apparently not designed with anything else than integer indices in mind. But that's exactly what we need here, so I use `__index__` directly as a workaround.

mattjj replied:
Can you leave a comment to this effect?

shoyer replied:
I don't think it makes sense to implement `__index__` on Poly then, since this violates Python's contract for the special method. In theory this could cause trouble if somehow the Poly class leaks into user code. I would just write your own helper function, e.g.,
```
def my_index(x):
   return x if isinstance(x, Poly) else operator.index(x)
```

JuliusKunze replied:
@shoyer Good point. Extracted a custom method now. @mattjj Clarified the choice there.

juliuskunze committed:
Fix dynamic slicing

juliuskunze committed:
Fix issue with float64.__index__()

juliuskunze committed:
Fix np.arange with float size, _try_canonicalize_shape

JuliusKunze commented:
@mattjj We might want to break up the tests into 2 files (shapecheck and masking), maybe even the interpreter file. Thoughts?

mattjj commented:
Splitting files SGTM, whatever you think is best. 

Re attaching traces to Polys, we might have a lifetime problem: Polys can outlive the trace, embedded in jaxprs, but we don’t want to leak trace objects. We could fix that with weakrefs but having an explicitly managed data structure might be better.

juliuskunze committed:
Cleanup: Make methods to create Poly internal (only use in Poly / shape spec parsing)

juliuskunze committed:
Fix testReshapeWithUnusualShapes (error message)

juliuskunze committed:
Fix syntax for python 3.6

juliuskunze committed:
Remove Poly.__index__

juliuskunze committed:
Merge branch 'master' of https://github.com/google/jax into shapecheck-pcnn

juliuskunze committed:
Fix tests

juliuskunze committed:
Split up masking.py

juliuskunze committed:
Cleanup masking

juliuskunze committed:
Cleanup

juliuskunze committed:
Use abstract_eval for shapecheck, remove ShapeCheckTrace(r)

juliuskunze committed:
Remove shape_rules, fix test

juliuskunze committed:
Remove shapes.py, move code to abstract_arrays.py / api.py

JuliusKunze commented:
@j-towns and I found that `shapecheck` does not require a custom trace(r). Instead, we can reuse `abstract_eval_fun` to propagate polymorphic shapes. This works since polymorphic shape evaluation is now based on the default shape rules. For this reason, I removed `ShapeCheckTrace(r)` and the `shape_rules` dictionary.

I also moved all remaining code needed by `shapecheck` out of `masking.py`:
- I moved `Poly` and related code (~160 lines) into `abstract_arrays.py`, since any code handling abstract arrays should also be able to handle polymorphic shapes now. It also helped to avoid a cyclic dependency problem [here](https://github.com/google/jax/pull/2017/files#diff-aa09bb80f1bb19fb0a4bac45f0e65b6eR164).
- The shape spec parser (~70 lines) is now in `api.py`, where it is used for `shapecheck` and `mask`.

@mattjj @shoyer Let me know if you know a better location for these.

mattjj commented on jax/abstract_arrays.py:
```diff
@@ -16,12 +16,18 @@
 from . import dtypes
-from . util import prod, partialmethod
+from . util import prod, partialmethod, safe_map, safe_zip
+map = safe_map
```
The code in this file is used so much that it's actually performance-critical; `safe_map`/`safe_zip` are too to slow to use here. Without looking at benchmarks to be sure, I'd say avoid these aliases (i.e. the `map = safe_map` and `zip = safe_zip` lines) and call `safe_map` / `safe_zip` explicitly in this file if/when needed.

JuliusKunze replied:
Ok, done!

mattjj commented on jax/abstract_arrays.py:
```diff
@@ -154,7 +160,8 @@ def __init__(self, val, weak_type=False):
     self.val = val
-    assert self.dtype != onp.dtype('O')
+
+    assert self.dtype != onp.dtype('O') or is_polymorphic(val)
```
I'd rather avoid storing object-dtype arrays, even if we create them as intermediates (by making `onp` operate on tuples of our Poly instances).

That said, I don't see why we'd need to create object-dtype arrays of _values_, just of shapes. So why would this check be needed?

mattjj replied:
Also, the definition of `is_polymorphic` looks like it's meant to apply to shapes, not values, but here it's being applied to a value. I'm confused!

JuliusKunze replied:
As I understand it, the argument for `gather`'s `start_indices` can an abstract value, i. e. when  jitting dynamic indexing. On the other hand, it can be a polymorphic value, i. e. `'n-1'` when indexing `x[-1]` where `x` is of polymorphic shape `'n'`. [This line](https://github.com/google/jax/pull/2017/files#diff-08adae3ce53df47932e06971edb68642R3030) reveals more of what's going on. This is the case where we need this atm AFAIK. Let me know if you have an idea how to avoid this.

mattjj replied:
The abstract value in that case should be a ShapedArray of dtype int, right? A ConcreteArray represents a specific array value, but we have no concrete value here.

I think we need to add abstraction handlers for Poly instances, which will produce something at the Shaped level unless they represent constant polynomials. We don't want to treat them as concrete arrays of dtype object; that'd be mixing in a new way of representing sets of possible values, but we already have a way to do that.

JuliusKunze replied:
@mattjj That makes sense. I think this would require to first solve [the problem we discussed earlier](https://github.com/google/jax/pull/2017#issue-363890835) though: `idx` is passed into `index_to_gather` without being wrapped into a trace for cases like `x[-1]`, resulting in the polymorphic value `n-1` for `i` [here](https://github.com/google/jax/pull/2017/files#diff-a4404c86bdd5bdfd0128f72b7aa1d40bR2965). Representing `i` as `ShapedArray((), int32)` would therefore currently break further processing due to abstract values getting passed into implementations. Lifting `i` back into a trace should allow getting rid of the current workaround.

JuliusKunze replied:
**Update:** I found a workaround for `gather` that is much simpler and doesn't require trace lifting. It allowed me to retighten the assertion above to the state before and remove other special cases, further minimizing the diff.

mattjj commented on jax/abstract_arrays.py:
```diff
@@ -248,3 +255,167 @@ def _make_concrete_python_scalar(x):
+
+# TODO remove remaining usages:
+def is_polymorphic(shape):
+  return any(map(lambda d: type(d) is Poly, shape))
```
`isindex` is kinda slow, so unless you think inheritance may be present (which IIUC it shouldn't be here), I think it's better to use `type(x) is Poly`.

JuliusKunze replied:
Sure, done!

mattjj commented on jax/abstract_arrays.py:
```diff
@@ -248,3 +255,167 @@ def _make_concrete_python_scalar(x):
+       for k, coeff in self.items()}), self.get(Mon(), 0) % divisor
+
+  def __hash__(self):
+    return hash(super())
```
Is this definition not just inherited? (Are dicts hashable?)

mattjj commented on jax/abstract_arrays.py:
```diff
@@ -248,3 +255,167 @@ def _make_concrete_python_scalar(x):
+    raise ValueError('Polynomials comparison "{} >= {}" is inconclusive.'
+                     .format(self, other))
+
+  def __le__(self, other):
```
You might be able to use `functools.total_ordering` instead of writing out all these explicitly.

JuliusKunze replied:
@mattjj Tried that: It breaks comparison with integers (needed for many shape rules).

mattjj commented on jax/api.py:
```diff
@@ -52,17 +53,11 @@
+                             local_devices, host_id, host_ids, host_count)
+from .abstract_arrays import ConcreteArray, ShapedArray, raise_to_shaped, \
+  eval_polymorphic_shape, Poly, Mon
+from .interpreters import partial_eval as pe, xla, pxla, ad, batching, parallel
```
Can you avoid changes to unrelated import statements?

JuliusKunze replied:
Sure. I reverted import changes to a minimum.

mattjj commented:
Overall that reorganization sounds reasonable, but you need to split this up into smaller changes to make it easier to review. For example, one PR could just be moving code without adding any features / methods etc. Other PRs could be adding specific features. (Also, none of the PRs should make unrelated changes, like rewriting import statements for unrelated code, or changing existing functions to use `safe_map` instead of `map`, etc.)

WDYT?

juliuskunze committed:
Remove safe_map/zip, is_instance from abstract_arrays, test + fix Poly hash, minimize import diff

juliuskunze committed:
Add missing shapecheck_test.py

juliuskunze committed:
Cleanup, minimize changes

juliuskunze committed:
Minimize import diff

juliuskunze committed:
Minor

JuliusKunze commented:
@mattjj I have minimized the diff quite a lot, and addressed all issues.

juliuskunze committed:
Allow shapecheck of np.where

juliuskunze committed:
Fix np.where

juliuskunze committed:
Simplify gather to allow retightening type assertion in ConcreteArray

juliuskunze committed:
Remove unused imports

j-towns commented on jax/api.py:
```diff
@@ -54,15 +54,15 @@
                              host_id, host_ids, host_count)
-from .abstract_arrays import ConcreteArray, ShapedArray, raise_to_shaped
+from .abstract_arrays import ConcreteArray, ShapedArray, raise_to_shaped, \
+  eval_polymorphic_shape, Poly, Mon
```
Style here is inconsistent with the line above.

j-towns commented on jax/numpy/lax_numpy.py:
```diff
@@ -45,7 +45,8 @@
 from .. import dtypes
-from ..abstract_arrays import UnshapedArray, ShapedArray, ConcreteArray
+from ..abstract_arrays import UnshapedArray, ShapedArray, ConcreteArray, \
+  is_polymorphic, to_index, Poly
```
Same here, use brackets rather than \

juliuskunze committed:
Make import style consistent

j-towns commented:
I don't think it's sensible to merge hacky workarounds like [this](https://github.com/google/jax/pull/2017/files#diff-a4404c86bdd5bdfd0128f72b7aa1d40bR1066-R1071), [this](https://github.com/google/jax/pull/2017/files#diff-a4404c86bdd5bdfd0128f72b7aa1d40bR1066-R1071) and [this](https://github.com/google/jax/pull/2017/files#diff-7628a7bafb266b297fa2216e22b600faR274-R280) into master.

@JuliusKunze if we urgently need the functionality we can setup our own branch and develop there. Merging into JAX master should wait until we have a cleaner and more general solution like the one discussed above.

juliuskunze committed:
Remove is_polymorphic, special cases in sampling, split, where.

mattjj commented:
Thanks for continuing to iterate and simplify!

Still I think we should split this into multiple _minimal_ PRs, where by minimal I mean smallest change unit. That makes it easier to review (reducing the probability we introduce bugs), easier to find the provenance of a bug later (using git bisect) and easier to roll any parts back if we need. This is something I’m trying to get better at myself :)

WDYT? What are the minimal units here?

juliuskunze committed:
Move back Poly, _parse_shape_spec into masking.py to simplify diff

juliuskunze committed:
Move back ShapeTest into masking_test.py to simplify diff

juliuskunze committed:
Minor reverts to further simplify diff

juliuskunze committed:
Fix tests

JuliusKunze commented:
@mattjj As discussed, I reverted the relocation for now to simplify the diff. I will create a separate PR with these changes once this one is merged.

JuliusKunze commented:
@mattjj I got tons of improvements in the pipeline, including shapecheck of `zeros`, `broadcast`, `split` ... (now works without hacks, I implemented the lifting) and masking rules for `conv`, `gather`, ... They all depend on this PR + the relocation, so I'll wait with PRs until this is merged.

juliuskunze committed:
Merge branch 'master' of https://github.com/google/jax into shapecheck-pcnn

# Conflicts:
#	jax/interpreters/masking.py

juliuskunze committed:
Merge remote-tracking branch 'main/master' into shapecheck-pcnn

# Conflicts:
#	jax/abstract_arrays.py
#	jax/interpreters/masking.py

juliuskunze committed:
Minimize diff

mattjj commented on jax/interpreters/masking.py:
```diff
@@ -1,74 +1,57 @@
-# Copyright 2019 Google LLC
```
Accidentally deleted the copyright notice?

mattjj replied:
I wanted to run Google-internal tests before merging, to make sure we don't break anything, but it turns out I can't sync the code down to Google until the copyright message is restored!

JuliusKunze replied:
Now fixed, was lost while moving around code between files and back.

juliuskunze committed:
Restore copyright, cleanup imports in masking.py

JuliusKunze commented:
@mattjj I got `jit(vmap(mask(pixelcnn)))` to work. I can prepare a PR as soon as this is merged. Is there anything holding back a merge?

juliuskunze committed:
Merge branch 'master' of https://github.com/google/jax into shapecheck-pcnn

# Conflicts:
#	jax/lax/lax.py

juliuskunze committed:
Merge branch 'master' of https://github.com/google/jax into shapecheck-pcnn

# Conflicts:
#	jax/api.py
#	jax/numpy/lax_numpy.py

juliuskunze committed:
Merge branch 'master' of https://github.com/google/jax into shapecheck-pcnn

# Conflicts:
#	jax/api.py
#	jax/numpy/lax_numpy.py

mattjj commented:
Nothing holding us back other than me being slow! Thanks for your patience, and for this *amazing* contribution.

mattjj reviewed (approved):
Woohoo, so much progress!

gnecula commented:
I am reverting this commit until we figure out how to handle this (a user complained, and there may be some other test failures).
